File: src/main/java/io/github/shaksternano/borgar/media/io/reader/FFmpegMediaReader.java
package io.github.shaksternano.borgar.media.io.reader;

import io.github.shaksternano.borgar.io.FileUtil;
import io.github.shaksternano.borgar.media.VideoFrame;
import io.github.shaksternano.borgar.util.MiscUtil;
import io.github.shaksternano.borgar.util.collect.ClosableIterator;
import org.apache.commons.io.FileUtils;
import org.bytedeco.javacv.FFmpegFrameGrabber;
import org.bytedeco.javacv.Frame;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

public abstract sealed class FFmpegMediaReader<E extends VideoFrame<?, E>> extends BaseMediaReader<E> permits FFmpegImageReader, FFmpegAudioReader {

    private final File input;
    protected final FFmpegFrameGrabber grabber;
    protected final List<AutoCloseable> toClose = new ArrayList<>();
    private final boolean deleteFile;
    private boolean closed = false;

    public FFmpegMediaReader(File input, String format) throws IOException {
        this(input, format, false);
    }

    public FFmpegMediaReader(InputStream input, String format) throws IOException {
        this(writeToFile(input, format), format, true);
    }

    private FFmpegMediaReader(File input, String format, boolean deleteFile) throws IOException {
        super(format);
        this.input = input;
        this.deleteFile = deleteFile;
        this.grabber = createGrabber();
        toClose.add(grabber);
        grabber.start();
        int frameCount = 0;
        Frame frame;
        while ((frame = grabFrame(grabber)) != null) {
            frameCount++;
            frame.close();
        }
        frameRate = grabber.getFrameRate();
        this.frameCount = frameCount;
        duration = grabber.getTimestamp();
        frameDuration = 1_000_000 / frameRate;
        audioChannels = grabber.getAudioChannels();
        audioSampleRate = grabber.getSampleRate();
        audioBitrate = grabber.getAudioBitrate();
        width = grabber.getImageWidth();
        height = grabber.getImageHeight();
    }

    private static File writeToFile(InputStream input, String format) throws IOException {
        var file = FileUtil.createTempFile("video", format);
        FileUtils.copyInputStreamToFile(input, file);
        return file;
    }

    protected FFmpegFrameGrabber createGrabber() {
        return new FFmpegFrameGrabber(input);
    }

    @Override
    public E readFrame(long timestamp) throws IOException {
        long circularTimestamp = timestamp % Math.max(duration(), 1);
        return frameNonCircular(circularTimestamp, grabber);
    }

    protected E frameAtTime(long timestamp, FFmpegFrameGrabber grabber) throws IOException {
        long circularTimestamp = timestamp % Math.max(duration() + 1, 1);
        return frameNonCircular(circularTimestamp, grabber);
    }

    @Override
    public E first() throws IOException {
        return frameNonCircular(0, grabber);
    }

    private E frameNonCircular(long timestamp, FFmpegFrameGrabber grabber) throws IOException {
        var frame = findFrame(timestamp, grabber);
        return convertFrame(frame);
    }

    private Frame findFrame(long timestamp, FFmpegFrameGrabber grabber) throws IOException {
        setTimestamp(timestamp, grabber);
        var frame = grabFrame(grabber);
        if (frame == null) {
            throw new NoSuchElementException("No frame at timestamp " + timestamp);
        }
        // The next call to grabFrame() will overwrite the current frame object, so we need to clone it
        var correctFrame = frame.clone();
        // The frame grabbed might not have the exact timestamp, even if there is a frame with that timestamp.
        // We keep grabbing frames, until we find one with a timestamp greater or equal to the requested timestamp.
        while (correctFrame.timestamp < timestamp) {
            var newFrame = grabFrame(grabber);
            if (newFrame == null || newFrame.timestamp > timestamp) {
                return correctFrame;
            }
            correctFrame.close();
            var copy = newFrame.clone();
            newFrame.close();
            correctFrame = copy;
        }
        return correctFrame;
    }

    protected abstract void setTimestamp(long timestamp, FFmpegFrameGrabber grabber) throws IOException;

    @Nullable
    protected abstract Frame grabFrame(FFmpegFrameGrabber grabber) throws IOException;

    @Nullable
    protected E grabConvertedFrame(FFmpegFrameGrabber grabber) throws IOException {
        var frame = grabFrame(grabber);
        if (frame == null) {
            return null;
        }
        return convertFrame(frame);
    }

    protected abstract E convertFrame(Frame frame);

    @Override
    public ClosableIterator<E> iterator() {
        try {
            return new Iterator();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    @Override
    public void close() throws IOException {
        if (closed) {
            return;
        }
        closed = true;
        MiscUtil.closeAll(toClose);
        if (deleteFile) {
            input.delete();
        }
    }

    private class Iterator implements ClosableIterator<E> {

        private final FFmpegFrameGrabber grabber;
        @Nullable
        private E nextFrame;
        private boolean closed = false;

        private Iterator() throws IOException {
            grabber = createGrabber();
            grabber.start();
            nextFrame = grabConvertedFrame(grabber);
        }

        @Override
        public boolean hasNext() {
            return nextFrame != null;
        }

        @Override
        public E next() {
            if (nextFrame == null) {
                throw new NoSuchElementException();
            }
            E frame = nextFrame;
            try {
                nextFrame = grabConvertedFrame(grabber);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
            return frame;
        }

        @Override
        public void close() throws IOException {
            if (!closed) {
                closed = true;
                grabber.close();
            }
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/reader/FFmpegMediaReader.kt
package io.github.shaksternano.borgar.core.media.reader

import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.io.SuspendCloseable
import io.github.shaksternano.borgar.core.io.closeAll
import io.github.shaksternano.borgar.core.io.use
import io.github.shaksternano.borgar.core.media.VideoFrame
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.withContext
import org.bytedeco.javacv.FFmpegFrameGrabber
import org.bytedeco.javacv.Frame
import java.nio.file.Path
import kotlin.io.path.deleteIfExists
import kotlin.math.max
import kotlin.time.Duration
import kotlin.time.Duration.Companion.microseconds
import kotlin.time.Duration.Companion.seconds

abstract class FFmpegMediaReader<T : VideoFrame<*>>(
    protected val input: Path,
    private val isTempFile: Boolean,
    private val grabber: FFmpegFrameGrabber,
) : BaseMediaReader<T>() {

    final override val loopCount: Int = 0

    override suspend fun readFrame(timestamp: Duration): T = readFrame(timestamp, grabber)

    protected suspend fun readFrame(timestamp: Duration, grabber: FFmpegFrameGrabber): T {
        val circularTimestamp =
            if (timestamp == duration) timestamp
            else (timestamp.inWholeMicroseconds % max(duration.inWholeMicroseconds, 1)).microseconds
        return readFrameNonCircular(circularTimestamp, grabber)
    }

    private suspend fun readFrameNonCircular(timestamp: Duration, grabber: FFmpegFrameGrabber): T {
        val frame = findFrame(timestamp, grabber)
        return convertFrame(frame)
    }

    private suspend fun findFrame(timestamp: Duration, grabber: FFmpegFrameGrabber): Frame {
        setTimestamp(timestamp, grabber)
        val frame = grabFrame(grabber)
            ?: throw NoSuchElementException("No frame at timestamp $timestamp")
        // The next call to grabFrame() will overwrite the current frame object, so we need to clone it
        var correctFrame = frame.clone()
        // The frame grabbed might not have the exact timestamp, even if there is a frame with that timestamp.
        // We keep grabbing frames, until we find one with a timestamp greater or equal to the requested timestamp.
        while (correctFrame.timestamp.microseconds < timestamp) {
            val newFrame = grabFrame(grabber)
            if (newFrame == null || newFrame.timestamp.microseconds > timestamp) {
                return correctFrame
            }
            withContext(Dispatchers.IO) {
                correctFrame.close()
            }
            val copy = newFrame.clone()
            withContext(Dispatchers.IO) {
                newFrame.close()
            }
            correctFrame = copy
        }
        return correctFrame
    }

    protected abstract suspend fun setTimestamp(timestamp: Duration, grabber: FFmpegFrameGrabber)

    protected abstract suspend fun grabFrame(grabber: FFmpegFrameGrabber): Frame?

    private suspend fun grabConvertedFrame(grabber: FFmpegFrameGrabber): T? {
        val frame = grabFrame(grabber) ?: return null
        return convertFrame(frame)
    }

    protected abstract fun convertFrame(frame: Frame): T

    override suspend fun close() = closeAll(
        SuspendCloseable.fromBlocking(grabber),
        SuspendCloseable.fromBlocking {
            if (isTempFile) {
                input.deleteIfExists()
            }
        }
    )

    override fun asFlow(): Flow<T> = flow {
        val grabber = FFmpegFrameGrabber(input.toFile())
        withContext(Dispatchers.IO) {
            grabber.start()
        }
        SuspendCloseable.fromBlocking(grabber).use {
            var nextFrame = grabConvertedFrame(grabber)
            while (nextFrame != null) {
                emit(nextFrame)
                nextFrame = grabConvertedFrame(grabber)
            }
        }
    }
}

suspend fun <T : FFmpegMediaReader<*>> createReader(
    input: DataSource,
    factory: (
        input: Path,
        isTempFile: Boolean,
        grabber: FFmpegFrameGrabber,
        frameCount: Int,
        frameRate: Double,
        frameDuration: Duration,
    ) -> T
): T {
    val isTempFile = input.path == null
    val path = input.getOrWriteFile().path
    val grabber = FFmpegFrameGrabber(path.toFile())
    return withContext(Dispatchers.IO) {
        grabber.start()
        var frameCount = 0
        var frame: Frame?
        while (grabber.grabImage().also { frame = it } != null) {
            frameCount++
            frame?.close()
        }
        val frameRate = grabber.frameRate
        factory(
            path,
            isTempFile,
            grabber,
            frameCount,
            frameRate,
            (1 / frameRate).seconds,
        )
    }
}
